tango/integrations/beaker/executor.py

```diff
@@ -0,0 +1,575 @@
+    """
+    The default image. Used if neither ``beaker_image`` nor ``docker_image`` are set.
+    """
+
+    def __init__(
+        self,
+        workspace: Workspace,
+        clusters: List[str],
+        include_package: Optional[Sequence[str]] = None,
+        beaker_workspace: Optional[str] = None,
+        github_token: Optional[str] = None,
+        beaker_image: Optional[str] = None,
+        docker_image: Optional[str] = None,
+        datasets: Optional[List[DataMount]] = None,
+        env_vars: Optional[List[EnvVar]] = None,
+        venv_name: Optional[str] = None,
+        parallelism: Optional[int] = -1,
+        **kwargs,
+    ):
+        # Pre-validate arguments.
+        if beaker_image is None and docker_image is None:
+            beaker_image = self.DEFAULT_BEAKER_IMAGE
+        elif (beaker_image is None) == (docker_image is None):
+            raise ConfigurationError(
+                "Either 'beaker_image' or 'docker_image' must be specified for BeakerExecutor, but not both."
+            )
+
+        super().__init__(workspace, include_package=include_package, parallelism=parallelism)
+        self.beaker = Beaker.from_env(default_workspace=beaker_workspace, session=True, **kwargs)
+        self.beaker_image = beaker_image
+        self.docker_image = docker_image
+        self.datasets = datasets
+        self.env_vars = env_vars
+        self.clusters = clusters
+        self.venv_name = venv_name
+        self._is_cancelled = threading.Event()
+
+        try:
+            self.github_token: str = github_token or os.environ["GITHUB_TOKEN"]
+        except KeyError:
+            raise ConfigurationError(
+                "A GitHub personal access token with the 'repo' scope is required. "
+                "This can be set with the 'github_token' argument to the BeakerExecutor, "
+                "or as the environment variable 'GITHUB_TOKEN'."
+            )
+
+        # Ensure entrypoint dataset exists.
+        self._ensure_entrypoint_dataset()
+
+    def execute_step(self, step: Step) -> None:
+        raise NotImplementedError
+
+    def execute_step_graph(
+        self, step_graph: StepGraph, run_name: Optional[str] = None
+    ) -> ExecutorOutput:
+        import concurrent.futures
+
+        self._is_cancelled.clear()
+
+        # These will hold the final results which we'll update along the way.
+        successful: Set[str] = set()
+        failed: Set[str] = set()
+        not_run: Set[str] = set()
+
+        # Keeps track of steps that are next up to run on Beaker.
+        steps_to_run: Set[str] = set()
+        # These are steps that have been submitted to Beaker but haven't completed yet.
+        submitted_steps: Set[str] = set()
+        # Futures for tracking the Beaker runs for each step.
+        step_futures: List[concurrent.futures.Future] = []
+
+        uncacheable_leaf_steps = step_graph.uncacheable_leaf_steps()
+
+        def update_steps_to_run():
```

Comment:
Maybe we should shift some of this logic to some `utils`. Just so our logic for updating steps remains the same across `MulticoreExecutor` and `BeakerExecutor`.